core/src/main/java/io/crate/plugin/PluginLoader.java

```diff
@@ -0,0 +1,226 @@
+package io.crate.plugin;
+
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import io.crate.Plugin;
+import io.crate.core.CrateComponentLoader;
+import org.apache.xbean.finder.ResourceFinder;
+import org.elasticsearch.common.collect.MapBuilder;
+import org.elasticsearch.common.component.LifecycleComponent;
+import org.elasticsearch.common.inject.Module;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.settings.ImmutableSettings;
+import org.elasticsearch.common.settings.Settings;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class PluginLoader {
+
+    private static final String RESOURCE_PATH = "META-INF/services/";
+    private static PluginLoader INSTANCE;
+
+    private final Settings settings;
+    private final List<Plugin> plugins;
+    private final ImmutableMap<Plugin, List<CrateComponentLoader.OnModuleReference>> onModuleReferences;
+    private final ESLogger logger;
+
+    public static synchronized PluginLoader getInstance(Settings settings) {
+        if (INSTANCE == null) {
+            INSTANCE = new PluginLoader(settings);
+        }
+        return INSTANCE;
+    }
+
+    public PluginLoader(Settings settings) {
+        this.settings = settings;
+        logger = Loggers.getLogger(getClass().getPackage().getName(), settings);
+
+        ResourceFinder finder = new ResourceFinder(RESOURCE_PATH);
+        List<Class<? extends Plugin>> implementations = null;
+        try {
+            implementations = finder.findAllImplementations(Plugin.class);
+        } catch (ClassNotFoundException | IOException e) {
+            logger.trace("exception while searching for plugins", e);
+        }
+
+        if (implementations == null) {
+            plugins = ImmutableList.of();
+            onModuleReferences = ImmutableMap.of();
+            return;
+        }
+
+        MapBuilder<Plugin, List<CrateComponentLoader.OnModuleReference>> onModuleReferences = MapBuilder.newMapBuilder();
+        ImmutableList.Builder<Plugin> builder = ImmutableList.builder();
+        for (Class<? extends Plugin> pluginClass : implementations) {
+            Plugin plugin = loadPlugin(pluginClass);
+            builder.add(plugin);
+            List<CrateComponentLoader.OnModuleReference> onModuleReferenceList = loadModuleReferences(plugin);
+            if (!onModuleReferenceList.isEmpty()) {
+                onModuleReferences.put(plugin, onModuleReferenceList);
+            }
+        }
+        plugins = builder.build();
+        this.onModuleReferences = onModuleReferences.immutableMap();
+
+        logger.info("plugins loaded: {} ", Lists.transform(plugins, new Function<Plugin, String>() {
```

Comment:
check for ``logger.isInfoEnabled``  otherwise the Lists.transform stuff is executed always